zkevm-circuits/src/evm_circuit/execution/precompiles/ec_mul.rs

```diff
@@ -0,0 +1,519 @@
+use bus_mapping::precompile::{PrecompileAuxData, PrecompileCalls};
+use eth_types::{Field, ToLittleEndian, ToScalar, Word};
+use gadgets::util::{and, not, or, Expr};
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        step::ExecutionState,
+        util::{
+            common_gadget::RestoreContextGadget,
+            constraint_builder::{ConstrainBuilderCommon, EVMConstraintBuilder},
+            math_gadget::{IsZeroGadget, ModGadget},
+            CachedRegion, Cell, RandomLinearCombination,
+        },
+    },
+    table::CallContextFieldTag,
+    witness::{Block, Call, ExecStep, Transaction},
+};
+
+// Modulus for scalar field Fr
+const FR_N: [u8; 32] = [
+    0x01, 0x00, 0x00, 0xF0, 0x93, 0xF5, 0xE1, 0x43, 0x91, 0x70, 0xB9, 0x79, 0x48, 0xE8, 0x33, 0x28,
+    0x5D, 0x58, 0x81, 0x81, 0xB6, 0x45, 0x50, 0xB8, 0x29, 0xA0, 0x31, 0xE1, 0x72, 0x4E, 0x64, 0x30,
+];
+
+#[derive(Clone, Debug)]
+pub struct EcMulGadget<F> {
+    point_p_x_rlc: Cell<F>,
+    point_p_y_rlc: Cell<F>,
+    scalar_s_rlc: Cell<F>,
+    scalar_s_raw_rlc: Cell<F>,
+    point_r_x_rlc: Cell<F>,
+    point_r_y_rlc: Cell<F>,
+
+    p_x_is_zero: IsZeroGadget<F>,
+    p_y_is_zero: IsZeroGadget<F>,
+    s_is_zero: IsZeroGadget<F>,
+
+    // Two Words (s_raw, s_modded) that satisfies
+    // k * FR_N + s_modded = s_raw
+    // Used for proving correct modulo by Fr
+    s_mod_pair: (
+        RandomLinearCombination<F, 32>, // raw
+        RandomLinearCombination<F, 32>, // mod by FR_N
+    ),
+    n: RandomLinearCombination<F, 32>, // modulus
+    modword: ModGadget<F>,
+
+    is_success: Cell<F>,
+    callee_address: Cell<F>,
+    caller_id: Cell<F>,
+    call_data_offset: Cell<F>,
+    call_data_length: Cell<F>,
+    return_data_offset: Cell<F>,
+    return_data_length: Cell<F>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for EcMulGadget<F> {
+    const NAME: &'static str = "EC_MUL";
+    const EXECUTION_STATE: ExecutionState = ExecutionState::PrecompileBn256ScalarMul;
+
+    fn configure(cb: &mut EVMConstraintBuilder<F>) -> Self {
+        let (
+            point_p_x_rlc,
+            point_p_y_rlc,
+            scalar_s_rlc,
+            scalar_s_raw_rlc,
+            point_r_x_rlc,
+            point_r_y_rlc,
+        ) = (
+            cb.query_cell_phase2(),
+            cb.query_cell_phase2(),
+            cb.query_cell_phase2(),
+            cb.query_cell_phase2(),
+            cb.query_cell_phase2(),
+            cb.query_cell_phase2(),
+        );
+
+        let (s_raw, s_modded, n) = (
+            cb.query_word_rlc(),
+            cb.query_word_rlc(),
+            cb.query_word_rlc(),
```

Comment:
`query_word_rlc` is used to match the intermediate cells in the `ModGadget`. Switching to `query_keccak_rlc` will make otherwise valid intermediate checks within `ModGadget` fail.

Reply:
ok, just noticed it. That's because the `ModGadget` was created to be used only for EVM inputs, that are RLC'd with the `challenges.evm_word()` randomness.

In our case, if you replace the `Phase2` cells by these `RandomLinearCombination`s, the constraint in `PrecompileGadget` for `input_rlc` equality will fail. Because the input bytes are RLC'd with `challenges.keccak_input()` randomness (input bytes are read in the Copy Circuit).

The only way out I see is a refactoring of the `ModGadget` for now, to allow the use-case to define which randomness to use. In the future, as we discussed in https://github.com/scroll-tech/zkevm-circuits/pull/600#discussion_r1266713577 we will anyway have to move away from the `ModGadget` in favour of `ModByNGadget` (where `N` is not an EVM variable).

Reply:
Just pushed a commit to fix this part (its a minor refactoring). Now we can remove the `Phase2` cells and also accordingly refactor `PrecompileGadget`